concepts/type-theory-glossary.md

```diff
@@ -0,0 +1,396 @@
+
+(This is also a great example for how we can use cardinality counting methods
+to create clear comparisons between differently named but functionally
+identical concepts between programming language communities!)
+
+### product types in the wild
+
+In Java, `class` creates a product type.
+
+In Go, `struct` creates a product type.
+
+In Rust, `struct` creates a product type.
+
+### interfaces in the wild
+
+In Java, the `interface` keyword introduces an interface.
+These interfaces are [*explicit*](#explicit-versus-structural-interfaces) -- a class must use the `implements` keyword to explicitly state that it can satisfy that interface.
+(That the class has sufficient methods to match the interface is then also checked at compile time -- but having the methods is not necessary rather than sufficient for the "implements" relationship to be considered true).
+
+In Go, the `interface` keyword introduces an interface.
+These interfaces are [*structural*](#explicit-versus-structural-interfaces) -- a struct doesn't need to *say* it implements an interface;
+if it has the relevant methods, then you can start using values of that type wherever the interface is required.
+
+In Rust, the `trait` feature combined with the `impl`+`for` keywords is most comparable to the concept of interface in other languages.
+Implementing a trait is [*explicit*](#explicit-versus-structural-interfaces) -- a type must use the `impl`+`for` keywords to explicitly state the methods that satisfy the trait.
+
+### what about numbers?
+
+Numbers are... interesting.
+
+For the most part in computing, we tend to pretend and build abstractions as if
+numbers are [infinite scalars](#infinite-scalars).
+
+Of course, this is rarely true in practice.
+Integers are often implemented in such a way that values that would require beyond 32 (or 64, etc) bits are unrepresentable;
+arguably, this makes them more like a (very (*very*) large) enum than anything else.
+Floating point numbers implemented in IEE764 style also have their own entire host of foibles.
+
+The scope of this document is not to solve this conundrum, nor to try to convince you how you should consider numbers.
+However, in IPLD, we usually do consider integers to be [infinite scalars](#infinite-scalars),
+with the caveat that libraries processing data which exceeds the boundaries they can process should error and halt explicitly.
+(In IPLD, we have the good fortune that this definition is unusually easy to stick to:
+since we only carry data around, and don't provide mathematical operations on it,
+it's easy to expect that data which was processable on the way in will remain processable throughout the time we handle it.)
+
+
+Types in IPLD Schemas
+---------------------
+
+It probably won't surprise to you hear that IPLD Schema's core features
+map very closely to the categories of types we've discussed:
+
+- [enums](#enums) are `enum`.
+- [sum types](#sum-types) are `union`.
```

Comment:
This being the case, it's probably worth mentioning explicitly in the "sum types" section that C/C++ unions _are_ sum types but are...not really usable as such.

Reply:
Yeah, okay.  I've been trying to refrain from getting "too many" languages carving out an explicit mention, but this one is present enough in worldwide discourse that it's worth a mention.